71_RayTracingPipeline/app_resources/common.hlsl

```diff
@@ -0,0 +1,171 @@
+#ifndef RQG_COMMON_HLSL
+#define RQG_COMMON_HLSL
+
+#include "nbl/builtin/hlsl/cpp_compat.hlsl"
+
+NBL_CONSTEXPR uint32_t WorkgroupSize = 16;
+
+struct Material
+{
+	float32_t3 ambient;
+    float32_t3 diffuse;
+    float32_t3 specular;
+    float32_t shininess;
+    float32_t dissolve; // 1 == opaque; 0 == fully transparent
+    uint32_t illum; // illumination model (see http://www.fileformat.info/format/material/)
+};
+
+struct SProceduralGeomInfo
+{
+    float32_t3 center;
+    float32_t radius;
+    Material material;
+};
+
+struct Aabb
+{
+    float32_t3 minimum;
+    float32_t3 maximum;
+};
+
+struct STriangleGeomInfo
+{
+    uint64_t vertexBufferAddress;
+    uint64_t indexBufferAddress;
+
+    uint32_t vertexStride : 29;
+    uint32_t indexType : 2; // 16 bit, 32 bit or none
+    uint32_t smoothNormals : 1;	// flat for cube, rectangle, disk
+
+    uint32_t objType;
+
+    Material material;
+};
+
+enum E_GEOM_TYPE : uint16_t
+{
+    EGT_TRIANGLES,
+    EGT_PROCEDURAL,
+    EGT_COUNT
+};
+
+enum E_RAY_TYPE : uint16_t
+{
+    ERT_PRIMARY, // Ray shoot from camera
+    ERT_OCCLUSION,
+    ERT_COUNT
+};
+
+enum E_MISS_TYPE : uint16_t
+{
+    EMT_PRIMARY,
+    EMT_OCCLUSION,
+    EMT_COUNT
+};
+
+enum E_LIGHT_TYPE : uint16_t
+{
+    ELT_DIRECTIONAL,
+    ELT_POINT,
+    ELT_SPOT,
+    ELT_COUNT
+};
+
+struct Light
+{
+    float32_t3 direction;
+    float32_t3 position;
+    float32_t outerCutoff;
+    uint16_t type;
```

Comment:
Don't use 16-bit types in PushConstants. they require a feature to be enabled and no AMD GPU supports them afaik

Reply:
yep bitfields of uint32_t are much better